File: cpp/arrays/1726-tuple-with-same-product.cpp
#include <map>
#include <vector>

using namespace std;


int tupleSameProduct(vector<int>& nums) {
    int response = 0;
    map<int, int> prods;
    for(int i = 0; i < nums.size(); i++) {
        for (int j  = i+1; j < nums.size(); j++) {
            prods[nums[i] * nums[j]]++;
        }
    }
    
    for(const auto& prod: prods) {
        if (prod.second >= 2) response += (prod.second * (prod.second - 1)) / 2;
    }
    
    return response * 8;
}

File: rust/arrays/1726-tuple-with-same-product.rs
use std::collections::HashMap;

pub fn tuple_same_product(nums: Vec<i32>) -> i32 {
    let mut prods = HashMap::new();
    let mut count = 0;
    for i in 0..nums.len() {
        for j in i+1..nums.len() {
            let prod = nums[i]*nums[j];
            
            count += 8 * prods.get(&prod).unwrap_or(&0);
            *prods.entry(nums[i] * nums[j]).or_insert(0) += 1;
        }
    }

    count
}